sdk/python/feast/infra/offline_stores/bigquery.py

```diff
@@ -286,24 +286,24 @@ def to_arrow(self) -> pyarrow.Table:
         return self.client.query(self.query).to_arrow()
 
 
-def block_until_done(client, bq_job):
-    def _is_done(job_id):
-        return client.get_job(job_id).state in ["PENDING", "RUNNING"]
-
+def _block_until_done(client, bq_job):
     @retry(wait=wait_fixed(10), stop=stop_after_delay(1800), reraise=True)
     def _wait_until_done(job_id):
-        return _is_done(job_id)
+        if client.get_job(job_id).state in ["PENDING", "RUNNING"]:
+            raise Exception
```

Comment:
When reading the code, that makes no sense to raise an Exception everytime the state of the job is simply RUNNING ...

Reply:
Before, the code inside the `@retry` annotation was not raising an `Exception`. Looking at the [tenacity retry docs](https://tenacity.readthedocs.io/en/latest/), it seems like this is the minimum to be take advantage of `retry`, is it not? Here I'm interpreting it as retrying every 10 seconds (up to max of 30 mins) until the job is anything but `PENDING` or `RUNNING` (i.e. `DONE`, `SUCCESS`, or `FAILURE` - see [docs](https://cloud.google.com/bigquery/docs/managing-jobs) for more).

Reply:
If you have more expertise on the `@retry` logic, do please let me know!!

Reply:
> In the old case, wouldn't it just pass first time around when there's no Exception raised?

Normally if the output of the function is a boolean, this should make it stop or continue
So I don't think that it was still retrying even when the job was done without exception. We've been using this function for the last couple of days and we never waited 30 min 🙂

Reply:
Let me double check 🙂

Reply:
Ok so something weird here but it sounds like we should use a [Custom callback](https://tenacity.readthedocs.io/en/latest/#custom-callbacks) where we pass our `_is_done()` function. For some reason, I thought that this was the current behaviour implemented 🤔 

For example, it works well with the following piece of code
```python
from tenacity import retry, wait_fixed, stop_after_delay, retry_if_result
from datetime import datetime

def get_state() -> int:
    number = int(datetime.now().timestamp())
    print(number)
    return number

def _is_done(number: int) -> bool:
    return (number % 10) != 0

@retry(wait=wait_fixed(1), stop=stop_after_delay(10), retry=retry_if_result(_is_done), reraise=True)
def test_retry():
    return get_state()
```

<img width="161" alt="image" src="https://user-images.githubusercontent.com/18557047/125513386-e4a6e1c7-a8af-44ab-a3bd-e55a808feeb7.png">

Reply:
> Normally if the output of the function is a boolean, this should make it stop or continue

Oh interesting. Are there docs for this? I saw the part about custom callbacks [here](https://tenacity.readthedocs.io/en/latest/#custom-callbacks) where it seems like you have to specify `retry=retry_if_result(is_false)` so I assumed it's not a default behavior.

Reply:
Just saw we commented around the same time haha. Let me implement the custom callback!

Reply:
Actually, thinking more about it, do you think it would be cleaner? Doesn't seem like we really need the return value of `_is_done` (whether the state is in ["DONE", "SUCCESS"]) anymore in my implementation.

Reply:
Yeah let me have a second thought 👍

Reply:
* later today